### PR TITLE
Asktell/variables objectives

### PR DESCRIPTION
**Pull request by jlnav**
Introducing specifying `variables` and `objectives` to ask/tell generators, This basically supersedes specifying gen_specs.in or gen_specs.out from the user's perspective.

For many of our gens, specify a mapping of variables and objectives to "x" and "f":

`variables_mapping = {"x": ["core", "edge"], "f": ["energy"]}`

### CODE REVIEW

**Comment by jlnav** [comment]
Dramatically simplified by the user providing a mapping:

```python

    variables = {"core": [-3, 3], "edge": [-2, 2]}
    objectives = {"energy": "EXPLORE"}
    mapping = {"x": ["core", "edge"]}

    gen = UniformSample(variables, objectives, mapping)
```

and:

```python

    variables = {"core": [-3, 3], "edge": [-2, 2]}
    objectives = {"energy": "MINIMIZE"}
    variables_mapping = {"x": ["core", "edge"], "f": ["energy"]}

    my_APOSMM = APOSMM(
        variables=variables, objectives=objectives, gen_specs=gen_specs, variables_mapping=variables_mapping
    )
```

Output dictionaries contain the variables, the objective value can be the key in `objectives`, and the generator uses the provided mapping to repack variables/objectives into "x"/"f".

_This is much simpler than my previous attempt at figuring this out dynamically!_

File: libensemble/generators.py [hunk 1]
```diff
@@ -95,12 +102,45 @@ class LibensembleGenerator(Generator):
...
+        self.variables_mapping = kwargs.get("variables_mapping", {})
+
+        self._internal_variable = "x"  # need to figure these out dynamically
+        self._internal_objective = "f"
+
+        if self.variables:
+
+            self.n = len(self.variables)
+            # build our own lb and ub
+            if "lb" not in kwargs and "ub" not in kwargs:
```
Comment (shuds13) [inline libensemble/generators.py:129]: Have you tested if these are in kargs?

Comment (shuds13) [reply libensemble/generators.py:129]: If I forget to comment out my lb/ub line I get:
...persistent_aposmm.py", line 326, in update_local_H_after_receiving
    local_H[name][Work["libE_info"]["H_rows"]] = calc_in[name]
ValueError: shape mismatch: value array of shape (4,6) could not be broadcast to indexing result of shape (4,3)

Comment (shuds13) [reply libensemble/generators.py:129]: That was with our branches merged.

**Comment by shuds13** [comment]
With your branch merged into mine the APOSMM run completes, but I dont have any local optimisation runs happening.

I ran with:

    vars_std = {"x0": [mcr, 1.0-mcr], "x1": [-20.0, 20.0], "x2":[1.0, 20.0],
                "x_on_cube0": [0, 1.0], "x_on_cube1": [0, 1.0], "x_on_cube2":[0, 1.0],
                }

    aposmm = APOSMM(
        variables=vars_std,
        objectives={"f": "MINIMIZE"},  # SH is "MINIMIZE" being used in APOSMM?
        persis_info=persis_info,
        initial_sample_size=initial_sample,
        #initial_sample_size=100,
        localopt_method="LN_BOBYQA",
        rk_const=0.5 * ((gamma(1 + (n / 2)) * 5) ** (1 / n)) / sqrt(pi),
        run_max_eval= 100 * (n + 1),
        max_active_runs=12,
        dist_to_bound_multiple=0.5,
        #lb=np.array([param.lower_bound for param in varying_params]),
        #ub=np.array([param.upper_bound for param in varying_params]),
        hfun=get_x,
        components=1,
        ftol_rel=1e-1,
    )

**Comment by shuds13** [comment]
Also tried with this, though I think this is the default right.

        mapping = {"x": ["x0", "x1", "x2", "x_on_cube0", "x_on_cube1", "x_on_cube2"]},

**Review by shuds13 — APPROVED** [review]

**Pull request merged**